File: src/main/java/org/cascadebot/cascadebot/music/MusicHandler.java
/*
 * Copyright (c) 2019 CascadeBot. All rights reserved.
 * Licensed under the MIT license.
 */

package org.cascadebot.cascadebot.music;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.source.bandcamp.BandcampAudioSourceManager;
import com.sedmelluq.discord.lavaplayer.source.beam.BeamAudioSourceManager;
import com.sedmelluq.discord.lavaplayer.source.soundcloud.SoundCloudAudioSourceManager;
import com.sedmelluq.discord.lavaplayer.source.twitch.TwitchStreamAudioSourceManager;
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager;
import lavalink.client.io.jda.JdaLavalink;
import lavalink.client.io.jda.JdaLink;
import lombok.AllArgsConstructor;
import lombok.Getter;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.TextChannel;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;
import org.apache.commons.lang3.StringUtils;
import org.apache.http.client.config.CookieSpecs;
import org.apache.http.client.config.RequestConfig;
import org.cascadebot.cascadebot.CascadeBot;
import org.cascadebot.cascadebot.data.Config;
import org.cascadebot.cascadebot.data.language.Language;
import org.cascadebot.cascadebot.data.managers.GuildDataManager;
import org.cascadebot.cascadebot.events.PlayerListener;
import org.cascadebot.cascadebot.messaging.Messaging;
import org.cascadebot.cascadebot.utils.PasteUtils;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MusicHandler {

    @Getter
    private AudioPlayerManager playerManager = new DefaultAudioPlayerManager();

    private Pattern typePattern = Pattern.compile("youtube#([A-z]+)");

    private CascadeBot instance;

    @Getter
    private String youtubeSourceName;
    @Getter
    private String twitchSourceName;

    @Getter
    private Map<Long, CascadePlayer> players = new HashMap<>();

    public MusicHandler(CascadeBot instance) {
        this.instance = instance;
    }

    JsonParser musicJsonParser = new JsonParser();

    private JdaLavalink lavalink;
    @Getter
    private boolean lavalinkEnabled;

    public void buildMusic() {
        AudioSourceManagers.registerRemoteSources(playerManager);

        YoutubeAudioSourceManager youtubeAudioSourceManager = new YoutubeAudioSourceManager(false);
        youtubeAudioSourceManager.configureRequests(config -> RequestConfig.copy(config).setCookieSpec(CookieSpecs.IGNORE_COOKIES).setConnectTimeout(5000).build());
        youtubeSourceName = youtubeAudioSourceManager.getSourceName();

        playerManager.registerSourceManager(youtubeAudioSourceManager);

        TwitchStreamAudioSourceManager twitchAudioManager = new TwitchStreamAudioSourceManager();
        twitchSourceName = twitchAudioManager.getSourceName();

        playerManager.registerSourceManager(twitchAudioManager);

        playerManager.registerSourceManager(new BeamAudioSourceManager());
        playerManager.registerSourceManager(SoundCloudAudioSourceManager.createDefault());
        playerManager.registerSourceManager(new BandcampAudioSourceManager());

        if (Config.INS.getMusicNodes().size() > 0) {
            lavalink = new JdaLavalink(Config.INS.getBotId().toString(), Config.INS.getShardNum(), shardId -> instance.getShardManager().getShardById(shardId));
            for (MusicNode musicNode : Config.INS.getMusicNodes()) {
                lavalink.addNode(musicNode.uri, musicNode.password); //TODO give nodes a name
            }
            lavalinkEnabled = true;
        } else {
            lavalinkEnabled = false;
        }

    }

    public CascadePlayer getPlayer(long guildId) {
        return players.computeIfAbsent(guildId, id -> {
            Guild guild = CascadeBot.INS.getShardManager().getGuildById(id);
            if (guild != null) {
                return createPlayer(guild);
            } else {
                return null;
            }
        });
    }

    private CascadePlayer createPlayer(Guild guild) {
        CascadePlayer player;
        if (isLavalinkEnabled()) {
            JdaLink link = getLavaLink().getLink(guild);
            player = new CascadeLavalinkPlayer(link.getPlayer());
        } else {
            AudioPlayer aPlayer = createLavaplayerPlayer();
            player = new CascadeLavaplayerPlayer(aPlayer);
            guild.getAudioManager().setSendingHandler(new LavaPlayerAudioSendHandler(aPlayer));
        }
        player.setGuild(guild);
        player.addListener(new PlayerListener(player));
        return player;
    }

    public boolean removePlayer(long guildId) {
        return players.remove(guildId) != null;
    }

    public void purgeDisconnectedPlayers() {
        players.entrySet().removeIf(entry -> entry.getValue().getConnectedChannel() == null
                && !GuildDataManager.getGuildData(entry.getKey()).getGuildTier().hasFlag("music_stay")
        );
    }

    /**
     * Searches for a list of 5 tracks and if it errors send the error to the specified channel
     *
     * @param search  The string to search
     * @param channel The {@link TextChannel} to send any errors to
     * @return The list of tracks that where found
     */
    public void searchTracks(String search, TextChannel channel, Consumer<List<SearchResult>> searchResultConsumer) {
        if (StringUtils.isBlank(Config.INS.getYoutubeKey())) {
            // TODO: Some way to disable searching?
            return;
        }
        Request request = new Request.Builder().url("https://www.googleapis.com/youtube/v3/search?part=snippet&q=" + URLEncoder.encode(search, StandardCharsets.UTF_8) + "&key=" + URLEncoder.encode(Config.INS.getYoutubeKey(), StandardCharsets.UTF_8) + "&maxResults=5&type=video,playlist").build();
        CascadeBot.INS.getHttpClient().newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Messaging.sendExceptionMessage(channel, Language.i18n(channel.getGuild().getIdLong(), "music.handler.error_searching"), e);
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) {
                if (response.body() == null) {
                    Messaging.sendDangerMessage(channel, Language.i18n(channel.getGuild().getIdLong(), "music.handler.no_data_returned"));
                    return;
                }

                try {
                    if (!response.isSuccessful()) {
                        Messaging.sendDangerMessage(channel, Language.i18n(channel.getGuild().getIdLong(), "music.handler.query_unsuccessful", PasteUtils.paste(response.body().string())));
                        return;
                    }
                    List<SearchResult> searchResults = new ArrayList<>();
                    JsonObject json = musicJsonParser.parse(response.body().string()).getAsJsonObject();
                    JsonArray items = json.getAsJsonArray("items");
                    int i = 0;
                    for (JsonElement elm : items) {
                        i++;
                        if (i > 5) {
                            CascadeBot.LOGGER.warn(Language.i18n(channel.getGuild().getIdLong(), "music.handler.more_than_five"));
                            break;
                        }
                        JsonObject item = elm.getAsJsonObject();
                        JsonObject idElm = item.getAsJsonObject("id");
                        String type = idElm.get("kind").getAsString();
                        Matcher matcher = typePattern.matcher(type);
                        if (!matcher.matches()) {
                            break;
                        }
                        type = matcher.group(1);
                        String url = "";
                        SearchResultType searchResultType = null;
                        switch (type) {
                            case "playlist":
                                searchResultType = SearchResultType.PLAYLIST;
                                url = "https://www.youtube.com/playlist?list=" + URLEncoder.encode(idElm.get("playlistId").getAsString(), StandardCharsets.UTF_8);
                                break;
                            case "video":
                                searchResultType = SearchResultType.VIDEO;
                                url = "https://www.youtube.com/watch?v=" + URLEncoder.encode(idElm.get("videoId").getAsString(), StandardCharsets.UTF_8);
                                break;
                        }
                        JsonObject snippetElm = item.getAsJsonObject("snippet");
                        String title = snippetElm.get("title").getAsString();
                        searchResults.add(new SearchResult(searchResultType, url, title));
                    }
                    searchResultConsumer.accept(searchResults);
                } catch (IOException e) {
                    Messaging.sendExceptionMessage(channel, Language.i18n(channel.getGuild().getIdLong(), "music.handler.error_reading"), e);
                } catch (Exception e) {
                    Messaging.sendExceptionMessage(channel, Language.i18n(channel.getGuild().getIdLong(), "music.handler.error_processing"), e);
                }
            }
        });
    }

    private AudioPlayer createLavaplayerPlayer() {
        return playerManager.createPlayer();
    }

    public JdaLavalink getLavalink() {
        return lavalink;
    }

    @AllArgsConstructor
    @Getter
    public static class MusicNode {

        URI uri;
        String password;

        //TODO maybe add port option?

    }

    @AllArgsConstructor
    @Getter
    public static class SearchResult {

        private SearchResultType type;
        private String url;
        private String title;

    }

    public enum SearchResultType {
        VIDEO, PLAYLIST
    }

    public JdaLavalink getLavaLink() {
        return lavalink;
    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/music/MusicHandler.kt
package org.cascadebot.cascadebot.music

import com.google.gson.JsonParser
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.source.bandcamp.BandcampAudioSourceManager
import com.sedmelluq.discord.lavaplayer.source.beam.BeamAudioSourceManager
import com.sedmelluq.discord.lavaplayer.source.soundcloud.SoundCloudAudioSourceManager
import com.sedmelluq.discord.lavaplayer.source.twitch.TwitchStreamAudioSourceManager
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager
import lavalink.client.io.jda.JdaLavalink
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response
import org.apache.commons.lang3.StringUtils
import org.apache.http.client.config.CookieSpecs
import org.apache.http.client.config.RequestConfig
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.data.Config
import org.cascadebot.cascadebot.data.language.Language
import org.cascadebot.cascadebot.data.managers.GuildDataManager
import org.cascadebot.cascadebot.data.objects.Flag
import org.cascadebot.cascadebot.events.PlayerListener
import org.cascadebot.cascadebot.messaging.MessageType
import org.cascadebot.cascadebot.messaging.Messaging
import org.cascadebot.cascadebot.messaging.Messaging.sendExceptionMessage
import org.cascadebot.cascadebot.utils.PasteUtils
import java.io.IOException
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.ArrayList
import java.util.function.Consumer
import java.util.function.Function
import java.util.regex.Pattern

class MusicHandler {
    val playerManager: AudioPlayerManager = DefaultAudioPlayerManager()
    val typePattern = Pattern.compile("youtube#([A-z]+)")

    val youtubeSourceName: String
    val twitchSourceName: String

    val players: MutableMap<Long, CascadePlayer> = HashMap()
    var musicJsonParser = JsonParser()
    var lavaLink: JdaLavalink? = null
        private set

    var lavalinkEnabled = false

    init {
        AudioSourceManagers.registerRemoteSources(playerManager)

        val youtubeAudioSourceManager = YoutubeAudioSourceManager(false)
        youtubeAudioSourceManager.configureRequests { RequestConfig.copy(it).setCookieSpec(CookieSpecs.IGNORE_COOKIES).setConnectTimeout(5000).build() }
        youtubeSourceName = youtubeAudioSourceManager.sourceName
        playerManager.registerSourceManager(youtubeAudioSourceManager)

        val twitchAudioManager = TwitchStreamAudioSourceManager()
        twitchSourceName = twitchAudioManager.sourceName
        playerManager.registerSourceManager(twitchAudioManager)
        playerManager.registerSourceManager(BeamAudioSourceManager())
        playerManager.registerSourceManager(SoundCloudAudioSourceManager.createDefault())
        playerManager.registerSourceManager(BandcampAudioSourceManager())
        if (Config.INS.musicNodes.size > 0) {
            lavaLink = JdaLavalink(Config.INS.botId.toString(), Config.INS.shardNum) { CascadeBot.INS.shardManager.getShardById(it!!) }
            for (musicNode in Config.INS.musicNodes) {
                lavaLink!!.addNode(musicNode.uri, musicNode.password) //TODO give nodes a name
            }
            lavalinkEnabled = true
        } else {
            lavalinkEnabled = false
        }
    }

    fun getPlayer(guildId: Long): CascadePlayer? {
        if (players.containsKey(guildId)) {
            return players[guildId]
        }
        val guild = CascadeBot.INS.shardManager.getGuildById(guildId)
        return if (guild != null) createPlayer(guild) else null
    }

    private fun createPlayer(guild: Guild): CascadePlayer {
        val player: CascadePlayer
        if (lavalinkEnabled) {
            val link = lavaLink!!.getLink(guild)
            player = CascadeLavalinkPlayer(link.player)
        } else {
            val aPlayer = createLavaplayerPlayer()
            player = CascadeLavaplayerPlayer(aPlayer)
            guild.audioManager.sendingHandler = LavaPlayerAudioSendHandler(aPlayer)
        }
        player.guild = guild
        player.addListener(PlayerListener(player)) //TODO dispose of players after a while
        return player
    }

    fun removePlayer(guildId: Long): Boolean {
        return players.remove(guildId) != null
    }

    fun purgeDisconnectedPlayers() {
        // Removes all players that are not connected to a channel unless they have supported us on Patreon
        players.entries.removeIf { it.value.connectedChannel == null && !GuildDataManager.getGuildData(it.key).isFlagEnabled(Flag.MUSIC_SERVICES) }
    }

    /**
     * Searches for a list of 5 tracks and if it errors send the error to the specified channel
     *
     * @param search  The string to search
     * @param channel The [TextChannel] to send any errors to
     * @return The list of tracks that where found
     */
    fun searchTracks(search: String?, channel: TextChannel, searchResultConsumer: Consumer<List<SearchResult>?>) {
        if (StringUtils.isBlank(Config.INS.youtubeKey)) {
            // TODO: Some way to disable searching?
            return
        }
        val request = Request.Builder().url("https://www.googleapis.com/youtube/v3/search?part=snippet&q=" + URLEncoder.encode(search, StandardCharsets.UTF_8) + "&key=" + URLEncoder.encode(Config.INS.youtubeKey, StandardCharsets.UTF_8) + "&maxResults=5&type=video,playlist").build()
        CascadeBot.INS.httpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                sendExceptionMessage(channel, Language.i18n(channel.guild.idLong, "music.handler.error_searching"), e)
            }

            override fun onResponse(call: Call, response: Response) {
                if (response.body == null) {
                    Messaging.sendMessage(MessageType.DANGER, channel, Language.i18n(channel.guild.idLong, "music.handler.no_data_returned"))
                    return
                }
                try {
                    if (!response.isSuccessful) {
                        Messaging.sendMessage(MessageType.DANGER, channel, Language.i18n(channel.guild.idLong, "music.handler.query_unsuccessful", PasteUtils.paste(response.body!!.string())))
                        return
                    }
                    val searchResults: MutableList<SearchResult> = ArrayList()
                    val json = musicJsonParser.parse(response.body!!.string()).asJsonObject
                    val items = json.getAsJsonArray("items")
                    var i = 0
                    for (elm in items) {
                        i++
                        if (i > 5) {
                            CascadeBot.LOGGER.warn(Language.i18n(channel.guild.idLong, "music.handler.more_than_five"))
                            break
                        }
                        val item = elm.asJsonObject
                        val idElm = item.getAsJsonObject("id")
                        var type = idElm["kind"].asString
                        val matcher = typePattern.matcher(type)
                        if (!matcher.matches()) {
                            break
                        }
                        type = matcher.group(1)
                        var url = ""
                        var searchResultType: SearchResultType? = null
                        when (type) {
                            "playlist" -> {
                                searchResultType = SearchResultType.PLAYLIST
                                url = "https://www.youtube.com/playlist?list=" + URLEncoder.encode(idElm["playlistId"].asString, StandardCharsets.UTF_8)
                            }
                            "video" -> {
                                searchResultType = SearchResultType.VIDEO
                                url = "https://www.youtube.com/watch?v=" + URLEncoder.encode(idElm["videoId"].asString, StandardCharsets.UTF_8)
                            }
                        }
                        val snippetElm = item.getAsJsonObject("snippet")
                        val title = snippetElm["title"].asString
                        searchResults.add(SearchResult(searchResultType!!, url, title))
                    }
                    searchResultConsumer.accept(searchResults)
                } catch (e: IOException) {
                    sendExceptionMessage(channel, Language.i18n(channel.guild.idLong, "music.handler.error_reading"), e)
                } catch (e: Exception) {
                    sendExceptionMessage(channel, Language.i18n(channel.guild.idLong, "music.handler.error_processing"), e)
                }
            }
        })
    }

    private fun createLavaplayerPlayer(): AudioPlayer {
        return playerManager.createPlayer()
    }

    class MusicNode(val uri: URI, val password: String)

    class SearchResult(val type: SearchResultType, val url: String, val title: String)

    enum class SearchResultType {
        VIDEO, PLAYLIST
    }

}
